### PR TITLE
Add endpoint subcommand for istioctl proxy-config

### PR DESCRIPTION
**Pull request by fisherxu**
Now we can't get endpoint list, so add endpoint subcommand for istioctl proxy-config to get endpoints, support for json, summary, filter(Endpoint address/port, Cluster name).

**List summary:**
```
istioctl proxy-config endpoint details-v1-7495976ddf-9s2kr
ENDPOINT               CLUSTER
127.0.0.1:9080         inbound|9080||details.default.svc.cluster.local
172.17.0.10:9411       outbound|9411||zipkin.istio-system.svc.cluster.local
172.17.0.10:14267      outbound|14267||jaeger-collector.istio-system.svc.cluster.local
172.17.0.10:14268      outbound|14268||jaeger-collector.istio-system.svc.cluster.local
172.17.0.10:16686      outbound|16686||jaeger-query.istio-system.svc.cluster.local
172.17.0.10:16686      outbound|80||tracing.istio-system.svc.cluster.local
172.17.0.11:9090       outbound|9090||prometheus.istio-system.svc.cluster.local
172.17.0.12:8088       outbound|8088||servicegraph.istio-system.svc.cluster.local
172.17.0.13:80         outbound|80||istio-ingressgateway.istio-system.svc.cluster.local
172.17.0.13:443        outbound|443||istio-ingressgateway.istio-system.svc.cluster.local
172.17.0.13:8060       outbound|8060||istio-ingressgateway.istio-system.svc.cluster.local
172.17.0.13:15011      outbound|15011||istio-ingressgateway.istio-system.svc.cluster.local
172.17.0.13:15030      outbound|15030||istio-ingressgateway.istio-system.svc.cluster.local
172.17.0.13:15031      outbound|15031||istio-ingressgateway.istio-system.svc.cluster.local
```

**List summary with filter:**
```
istioctl proxy-config endpoint details-v1-7495976ddf-9s2kr --port=443
ENDPOINT            CLUSTER
172.17.0.13:443     outbound|443||istio-ingressgateway.istio-system.svc.cluster.local
172.17.0.19:443     outbound|443||istio-galley.istio-system.svc.cluster.local
172.17.0.4:443      outbound|443||istio-sidecar-injector.istio-system.svc.cluster.local
172.17.0.6:443      outbound|443||istio-egressgateway.istio-system.svc.cluster.local
```

**List with json format:**
```
istioctl proxy-config endpoint details-v1-7495976ddf-9s2kr --address=127.0.0.1 -ojson
[
    {
        "name": "inbound|9080||details.default.svc.cluster.local",
        "addedViaApi": true,
        "hostStatuses": [
            {
                "address": {
                    "socketAddress": {
                        "address": "127.0.0.1",
                        "portValue": 9080
                    }
                },
                "stats": {
                    "cx_active": {
                        "type": "GAUGE"
                    },
                    "cx_connect_fail": {},
                    "cx_total": {
                        "value": "8"
                    },
                    "rq_active": {
                        "type": "GAUGE"
                    },
                    "rq_error": {},
                    "rq_success": {
                        "value": "9"
                    },
                    "rq_timeout": {},
                    "rq_total": {
                        "value": "9"
                    }
                },
                "healthStatus": {
                    "edsHealthStatus": "HEALTHY"
                }
            }
        ]
    }
]
```

### CODE REVIEW

File: istioctl/cmd/istioctl/proxyconfig.go [hunk 1]
```diff
@@ -173,13 +174,57 @@ var (
...
+  istioctl proxy-config endpoint <pod-name>
+
+  # Retrieve endpoint summary for endpoint with port 9080.
+  istioctl proxy-config endpoint <pod-name> --port 9080
+
+  # Retrieve full endpoint with a address (172.17.0.2).
+  istioctl proxy-config endpoint <pod-name> --address 172.17.0.2 -o json
+
+  # Retrieve full endpoint with a cluster name (outbound|9411||zipkin.istio-system.svc.cluster.local).
+  istioctl proxy-config endpoint <pod-name> --name "outbound|9411||zipkin.istio-system.svc.cluster.local" -o json
```
Comment (liamawhite) [inline istioctl/cmd/istioctl/proxyconfig.go]: can we make this `--cluster` instead?

File: istioctl/cmd/istioctl/proxyconfig.go [hunk 2]
```diff
@@ -208,6 +253,24 @@ func setupEnvoyConfigWriter(podName, podNamespace string, out io.Writer) (*confi
 	return cw, nil
 }
 
+func setupClustersEnvoyConfigWriter(podName, podNamespace string, out io.Writer) (*clusters.ConfigWriter, error) {
```
Comment (liamawhite) [inline istioctl/cmd/istioctl/proxyconfig.go]: Could you add the following todo comment above this function:
```
// TODO: migrate this to config dump when implemented in Envoy
// Issue to track -> https://github.com/envoyproxy/envoy/issues/3362
```

File: istioctl/cmd/istioctl/proxyconfig.go [hunk 3]
```diff
@@ -223,5 +286,9 @@ func init() {
 
 	routeConfigCmd.PersistentFlags().StringVar(&routeName, "name", "", "Filter listeners by route name field")
 
-	configCmd.AddCommand(clusterConfigCmd, listenerConfigCmd, routeConfigCmd, bootstrapConfigCmd)
+	endpointConfigCmd.PersistentFlags().StringVar(&address, "address", "", "Filter endpoints by address field")
+	endpointConfigCmd.PersistentFlags().IntVar(&port, "port", 0, "Filter endpoints by Port field")
+	endpointConfigCmd.PersistentFlags().StringVar(&clusterName, "name", "", "Filter endpoints by cluster name field")
```
Comment (liamawhite) [inline istioctl/cmd/istioctl/proxyconfig.go]: Change to cluster as mentioned above ^^

File: istioctl/pkg/writer/envoy/clusters/clusters.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	clusters *clusters.Wrapper
+}
+
+type EndpointCluster struct {
+	address string
+	port    int
+	cluster string
+}
+
+// Prime loads the config dump into the writer ready for printing
```
Comment (liamawhite) [inline istioctl/pkg/writer/envoy/clusters/clusters.go]: `s/config dump/clusters output/`

File: istioctl/cmd/istioctl/proxyconfig.go (hunk 1 above)
Comment (fisherxu) [reply istioctl/cmd/istioctl/proxyconfig.go]: Done.

File: istioctl/cmd/istioctl/proxyconfig.go (hunk 2 above)
Comment (fisherxu) [reply istioctl/cmd/istioctl/proxyconfig.go]: have added, look forward to see eds in config_dump :)

File: istioctl/cmd/istioctl/proxyconfig.go (hunk 3 above)
Comment (fisherxu) [reply istioctl/cmd/istioctl/proxyconfig.go]: Done.

File: istioctl/pkg/writer/envoy/clusters/clusters.go (hunk 4 above)
Comment (fisherxu) [reply istioctl/pkg/writer/envoy/clusters/clusters.go]: Done.

File: istioctl/cmd/istioctl/proxyconfig.go [hunk 5]
```diff
@@ -173,13 +174,57 @@ var (
...
+  # Retrieve endpoint summary for endpoint with port 9080.
+  istioctl proxy-config endpoint <pod-name> --port 9080
+
+  # Retrieve full endpoint with a address (172.17.0.2).
+  istioctl proxy-config endpoint <pod-name> --address 172.17.0.2 -o json
+
+  # Retrieve full endpoint with a cluster name (outbound|9411||zipkin.istio-system.svc.cluster.local).
+  istioctl proxy-config endpoint <pod-name> --cluster "outbound|9411||zipkin.istio-system.svc.cluster.local" -o json
+`,
+		Aliases: []string{"ep"},
```
Comment (liamawhite) [inline istioctl/cmd/istioctl/proxyconfig.go]: Could you add endpoints as an alias as well?

**Comment by liamawhite** [comment]
Had just one more comment https://github.com/istio/istio/pull/8512/files#r215517206

File: istioctl/cmd/istioctl/proxyconfig.go (hunk 5 above)
Comment (fisherxu) [reply istioctl/cmd/istioctl/proxyconfig.go]: Sure, added.

**Comment by andraxylia** [comment]
@fisherxu Can you please cherry-pick this is release-1.0.0 and update the guide at 

https://istio.io/help/ops/traffic-management/proxy-cmd/

We need this for troubleshooting @hklai @costinm

**Comment by fisherxu** [comment]
@andraxylia Sure, I can cherry-pick it.

**Comment by fisherxu** [comment]
The PR #[2553](https://github.com/istio/istio.io/pull/2553) in istio.io has merged, but don't know why not display? @geeknoid @andraxylia

**Pull request merged**